File: src/main/java/dev/slne/protect/bukkit/listener/listeners/ProtectionModeListener.java
package dev.slne.protect.bukkit.listener.listeners;

import com.destroystokyo.paper.MaterialSetTag;
import dev.slne.protect.bukkit.PaperMain;
import dev.slne.protect.bukkit.region.ProtectionRegion;
import dev.slne.protect.bukkit.user.ProtectionUser;
import io.papermc.paper.event.player.PlayerItemFrameChangeEvent;
import io.papermc.paper.event.player.PrePlayerAttackEntityEvent;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.WorldBorder;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.EntityTargetEvent;
import org.bukkit.event.player.PlayerArmorStandManipulateEvent;
import org.bukkit.event.player.PlayerAttemptPickupItemEvent;
import org.bukkit.event.player.PlayerDropItemEvent;
import org.bukkit.event.player.PlayerInteractEntityEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerMoveEvent;
import org.bukkit.event.player.PlayerPortalEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.event.player.PlayerSwapHandItemsEvent;
import org.bukkit.event.server.PluginDisableEvent;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

public class ProtectionModeListener implements Listener {

  private static final MaterialSetTag INTERACTABLE = new MaterialSetTag(
      key("interactable")).contains("BUTTON").contains("DOOR").contains("LEVER").contains("GATE")
      .contains("TRAPDOOR").contains("ANVIL").contains("TABLE").contains("FURNACE")
      .contains("SMOKER").contains("BREWING_STAND").contains("LECTERN").contains("STONECUTTER")
      .contains("GRINDSTONE").contains("HOPPER").contains("DROPPER").contains("DISPENSER")
      .contains("CHEST").contains("SHULKER_BOX").contains("BARREL").contains("BEACON")
      .contains("BED").contains("CAULDRON").contains("COMPARATOR").contains("DAYLIGHT_DETECTOR")
      .contains("SIGN").contains("PRESSURE_PLATE").contains("JUKEBOX").contains("NOTE_BLOCK")
      .contains("BELL").contains("CAMPFIRE").contains("POT").add(Material.LOOM)
      .add(Material.REPEATER).add(Material.OBSERVER).add(Material.CHISELED_BOOKSHELF).lock();

  @Contract("_ -> new")
  private static @NotNull NamespacedKey key(String name) {
    return new NamespacedKey(PaperMain.getInstance(), name);
  }

  /**
   * Checks if the would interact instead of placing a block
   *
   * @param material the material
   * @return true if the material would be interacted with
   */
  private boolean isInteractive(Block block) {

    if (block == null) {
      return false;
    }

//    Material material = block.getType();
//    String name = material.toString();
//    Pattern pattern = Pattern.compile(
//        ".*(BUTTON|DOOR|LEVER|GATE|TRAPDOOR|ANVIL|TABLE|FURNACE|SMOKER|BREWING_STAND|LECTERN|STONECUTTER|LOOM|GRINDSTONE|HOPPER|DROPPER|DISPENSER|CHEST|"
//            + "SHULKER_BOX|BARREL|BEACON|BED|CAULDRON|COMPARATOR|DAYLIGHT_DETECTOR|REPEATER|OBSERVER|SIGN|PRESSURE_PLATE|JUKEBOX|NOTE_BLOCK|BELL|CAMPFIRE|POT|BOOKSHELF).*",
//        Pattern.CASE_INSENSITIVE);
//
//    return pattern.matcher(name).matches();

    return INTERACTABLE.isTagged(block);
  }

  @EventHandler(priority = EventPriority.LOWEST)
  public void onQuit(PlayerQuitEvent event) {
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(event.getPlayer());

    if (protectionUser.hasRegionCreation()) {
      protectionUser.getRegionCreation().cancelProtection();
      protectionUser.resetRegionCreation();
    }
  }

  @EventHandler
  public void onMove(PlayerMoveEvent event) {
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(event.getPlayer());
    ProtectionRegion regionCreation = protectionUser.getRegionCreation();

    if (regionCreation != null) {
      Location protectionModeLocation = regionCreation.getStartLocation();
      Player player = event.getPlayer();
      WorldBorder worldBorder = player.getWorldBorder();

      Location to = event.getTo();
      if (worldBorder != null && !worldBorder.isInside(to)) {
        event.getPlayer().teleport(protectionModeLocation);
      }

    }
  }

  @EventHandler
  public void onItemDrop(PlayerDropItemEvent event) {
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(event.getPlayer());

    if (protectionUser.hasRegionCreation()) {
      event.setCancelled(true);
    }
  }

  @EventHandler(priority = EventPriority.HIGHEST)
  public void onPortal(PlayerPortalEvent event) {
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(event.getPlayer());

    if (protectionUser.hasRegionCreation()) {
      event.setCancelled(true);
    }
  }

  @EventHandler
  public void onDamage(EntityDamageEvent event) {
    if (!(event.getEntity() instanceof Player player)) {
      return;
    }

    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(player);

    if (protectionUser.hasRegionCreation()) {
      event.setCancelled(true);
    }
  }

  @EventHandler
  public void onItemPickup(PlayerAttemptPickupItemEvent event) {
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(event.getPlayer());

    if (protectionUser.hasRegionCreation()) {
      event.setCancelled(true);
    }
  }

  @EventHandler
  public void onHandSwap(PlayerSwapHandItemsEvent event) {
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(event.getPlayer());

    if (protectionUser.hasRegionCreation()) {
      event.setCancelled(true);
    }
  }

  @EventHandler
  public void onEntityAttack(PrePlayerAttackEntityEvent event) {
    Player player = event.getPlayer();
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(player);

    if (protectionUser.hasRegionCreation()) {
      event.setCancelled(true);
    }
  }

  @EventHandler
  public void onEntityTarget(EntityTargetEvent event) {
    if (!(event.getTarget() instanceof Player player)) {
      return;
    }

    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(player);

    if (protectionUser.hasRegionCreation()) {
      event.setCancelled(true);
    }
  }

  @EventHandler
  public void onInteract(PlayerInteractEvent event) {
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(event.getPlayer());
    ProtectionRegion regionCreation = protectionUser.getRegionCreation();
    Block block = event.getClickedBlock();

    boolean isInProtectionMode = regionCreation != null;
    boolean isPlacingRedstoneTorch = event.getAction() == Action.RIGHT_CLICK_BLOCK
        && event.getMaterial() == Material.REDSTONE_TORCH;
    boolean isBreakingRedstoneTorch =
        event.getAction() == Action.LEFT_CLICK_BLOCK && (block.getType() == Material.REDSTONE_TORCH
            || block.getType() == Material.REDSTONE_WALL_TORCH);
    boolean isInteracting = isInteractive(block);

    // Check if the player is in protection mode else we don't want to do anything
    if (!isInProtectionMode) {
      return;
    }

    // Check if the player is interacting with a block instead of placing a redstone torch and if so cancel the event
    if (isInteracting && !event.getPlayer().isSneaking()) {
      event.setCancelled(true);
      return;
    }

    // Check if the player is placing or breaking a redstone torch and if not cancel the event
    if (!isPlacingRedstoneTorch && !isBreakingRedstoneTorch) {
      event.setCancelled(true);
    }

  }

  @EventHandler
  public void onItemFrameChange(PlayerItemFrameChangeEvent event) {
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(event.getPlayer());
    ProtectionRegion regionCreation = protectionUser.getRegionCreation();

    if (regionCreation != null) {
      event.setCancelled(true);
    }
  }

  @EventHandler
  public void onArmorstandManipulate(PlayerArmorStandManipulateEvent event) {
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(event.getPlayer());
    ProtectionRegion regionCreation = protectionUser.getRegionCreation();

    if (regionCreation != null) {
      event.setCancelled(true);
    }
  }

  @EventHandler
  public void onEntityRide(PlayerInteractEntityEvent event) {
    ProtectionUser protectionUser = ProtectionUser.getProtectionUser(event.getPlayer());
    ProtectionRegion regionCreation = protectionUser.getRegionCreation();

    if (regionCreation != null) {
      event.setCancelled(true);
    }
  }

  @EventHandler
  public void onServerStop(
      PluginDisableEvent event) { // TODO: 04.02.2024 11:10 - move to a better place
    for (ProtectionUser protectionUser : PaperMain.getBukkitInstance().getUserManager()
        .getUsers()) {
      if (protectionUser.hasRegionCreation()) {
        protectionUser.getRegionCreation().cancelProtection();
      }
    }
  }
}

File: src/main/kotlin/dev/slne/surf/protect/paper/listener/listeners/ProtectionModeListener.kt
package dev.slne.surf.protect.paper.listener.listeners

import com.destroystokyo.paper.MaterialSetTag
import dev.slne.surf.api.paper.util.namespacedKey
import dev.slne.surf.protect.paper.config
import dev.slne.surf.protect.paper.configManager
import dev.slne.surf.protect.paper.items.ProtectionItems
import dev.slne.surf.protect.paper.user.protectionUser
import io.papermc.paper.event.player.PlayerItemFrameChangeEvent
import io.papermc.paper.event.player.PrePlayerAttackEntityEvent
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.event.entity.EntityTargetEvent
import org.bukkit.event.player.*

object ProtectionModeListener : Listener {
    private fun Block.isInteractive() = INTERACTABLE.isTagged(this)

    @EventHandler(priority = EventPriority.LOWEST)
    fun onQuit(event: PlayerQuitEvent) {
        val player = event.player
        player.protectionUser().handleQuit(player)
    }

    @EventHandler
    fun onJoin(event: PlayerJoinEvent) {
        val player = event.player

        val awaitingProtectionMode =
            config.awaitingProtectionModes.firstOrNull { it.playerUuid == player.uniqueId }
                ?: return

        val protectionUser = player.protectionUser()

        protectionUser.restorePlayerProperties(player, awaitingProtectionMode.inventory)

        configManager.edit {
            awaitingProtectionModes.remove(awaitingProtectionMode)
        }

        player.teleportAsync(awaitingProtectionMode.startLocation)
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onMove(event: PlayerMoveEvent) {
        if (!(event.hasExplicitlyChangedBlock())) return
        val player = event.player
        val protectionUser = player.protectionUser()
        val regionCreation = protectionUser.regionCreation

        if (regionCreation != null) {
            val protectionModeLocation = regionCreation.startLocation
            val worldBorder = player.worldBorder

            if (worldBorder != null && !worldBorder.isInside(event.to)) {
                player.teleportAsync(protectionModeLocation)
            }
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onPlayerTeleport(event: PlayerTeleportEvent) {
        if (event.to.world != event.from.world && event.player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onItemDrop(event: PlayerDropItemEvent) {
        if (event.player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onPortal(event: PlayerPortalEvent) {
        if (event.player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onDamage(event: EntityDamageEvent) {
        val player = event.entity as? Player ?: return
        if (player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onItemPickup(event: PlayerAttemptPickupItemEvent) {
        if (event.player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onHandSwap(event: PlayerSwapHandItemsEvent) {
        if (event.player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onEntityAttack(event: PrePlayerAttackEntityEvent) {
        if (event.player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onEntityTarget(event: EntityTargetEvent) {
        val player = event.target as? Player ?: return
        if (player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onInteract(event: PlayerInteractEvent) {
        val player = event.player
        if (!player.protectionUser().isCreatingRegion) return
        val block = event.clickedBlock
        val isInteracting = block?.isInteractive() ?: false

        // Check if the player is interacting with a block instead of placing a redstone torch and if so cancel the event
        if (isInteracting && !player.isSneaking) {
            event.setCancelled(true)
            return
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onBlockBreak(event: BlockBreakEvent) {
        if (!event.player.protectionUser().isCreatingRegion) return
        if (!(ProtectionItems.isProtectionBlock(event.block))) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onBlockPlace(event: BlockPlaceEvent) {
        if (!event.player.protectionUser().isCreatingRegion) return
        if (!(ProtectionItems.isProtectionItem(event.itemInHand))) {
            event.isCancelled = true
        }
    }


    @EventHandler(priority = EventPriority.LOWEST)
    fun onItemFrameChange(event: PlayerItemFrameChangeEvent) {
        if (event.player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onArmorstandManipulate(event: PlayerArmorStandManipulateEvent) {
        if (event.player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    @EventHandler(priority = EventPriority.LOWEST)
    fun onEntityRide(event: PlayerInteractEntityEvent) {
        if (event.player.protectionUser().isCreatingRegion) {
            event.isCancelled = true
        }
    }

    private val INTERACTABLE = MaterialSetTag(namespacedKey("interactable"))
        .contains("BUTTON")
        .contains("DOOR")
        .contains("LEVER")
        .contains("GATE")
        .contains("TRAPDOOR")
        .contains("ANVIL")
        .contains("TABLE")
        .contains("FURNACE")
        .contains("SMOKER")
        .contains("BREWING_STAND")
        .contains("LECTERN")
        .contains("STONECUTTER")
        .contains("GRINDSTONE")
        .contains("HOPPER")
        .contains("DROPPER")
        .contains("DISPENSER")
        .contains("CHEST")
        .contains("SHULKER_BOX")
        .contains("BARREL")
        .contains("BEACON")
        .contains("BED").not(Material.BEDROCK) // BEDROCK is not a bed
        .contains("CAULDRON")
        .contains("COMPARATOR")
        .contains("DAYLIGHT_DETECTOR")
        .contains("SIGN")
        .contains("PRESSURE_PLATE")
        .contains("JUKEBOX")
        .contains("NOTE_BLOCK")
        .contains("BELL")
        .contains("CAMPFIRE")
        .contains("POT")
        .add(Material.LOOM)
        .add(Material.REPEATER)
        .add(Material.OBSERVER)
        .add(Material.CHISELED_BOOKSHELF)
        .add(Material.CRAFTER)
        .lock()
}
